### PR TITLE
New Teams resources

### PR DESCRIPTION
**Pull request by desmay**
#### Pull Request (PR) description
Adding new MS Teams resources.

#### Task list

- [x] Added an entry under the Unreleased section of the change log in the CHANGELOG.md.
      Entry should say what was changed, and how that affects users (if applicable).
- [x] Resource documentation added/updated in README.md in the resource folder.
- [x] Resource parameter descriptions added/updated in schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [x] Integration tests added/updated (where possible). See [DSC Resource Testing Guidelines](https://github.com/PowerShell/DscResources/blob/master/TestsGuidelines.md).
- [x] New/changed code adheres to [DSC Resource Style Guidelines](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md) and [Best Practices](https://github.com/PowerShell/DscResources/blob/master/BestPractices.md).

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/office365dsc/52)

### CODE REVIEW

**Comment by ykuijs** [comment]
There is one more test failing:
https://ci.appveyor.com/project/NikCharlebois/office365dsc/builds/22171212#L300

**Comment by desmay** [comment]
ykuijs I was chatting with Nik today, I'm gonna try to fix this test this evening after work..

**Comment by NikCharlebois** [comment]
@desmay you will need to declare the following in the TeamsTeam tests. Otherwise, the test tries to call the "real" Invoke-ExoCommand and tries to create a connection to exchange online using bogus credentials, which then throws an error.

Mock Invoke-ExoCommand {
            return Invoke-Command -ScriptBlock $ScriptBlock -ArgumentList $Arguments -NoNewScope
        }

**Comment by desmay** [comment]
@NikCharlebois Thank you for help again, I'm not sure how I missed that last night was comparing my test to other EXO tests.  Anyhow just made change and it fixed test

**Comment by desmay** [comment]
Looks like I need to add a few more tests

**Comment by NikCharlebois** [comment]
@desmay yes, look at the list of lines that are not covered at the bottom of the AppVeyor report. We need to achieve at least 90% coverage overall. Thanks

**Comment by desmay** [comment]
@ykuijs Yeah I used VS code format.. I'll see if there is a way to change format settings in VSCode

**Comment by ykuijs** [comment]
There sure is. Nik and I are already looking into that! :-)

**Comment by NikCharlebois** [comment]
Ask and you shall receive :) .vscode settings have been committed to the Dev main branch

**Comment by ykuijs** [comment]
One more comment: Also please update the Changelog.md file with your updates.

**Comment by RobBiddle** [comment]
@desmay this PR looks like a great addition!  I'm definitely looking forward to seeing it accepted as I'm in want of Teams related DSC for a current work project.  If you would like assistance banging keys I'd be glad to help.

File: Modules/Office365DSC/DSCResources/MSFT_TeamsChannel/MSFT_TeamsChannel.psm1 [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+
+    if ($Ensure -eq "Present")
+    {
+        # Remap attribute from DisplayName to current display name for Set-TeamChannel cmdlet
+        if ($channel.Ensure -eq "Present")
+        {
+            if ($CurrentParameters.ContainsKey("NewDisplayName"))
+            {
+                $CurrentParameters.Add("CurrentDisplayName", $DisplayName)
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsChannel/MSFT_TeamsChannel.psm1]: Sounds good, I wish cmdlets for Add-Channel / Set-Channel both used same name displaychannel for consistency sake.

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+        [Parameter(Mandatory = $true)]
+        [System.Management.Automation.PSCredential]
+        $GlobalAdminAccount
+    )
+
+    Test-TeamsServiceConnection -GlobalAdminAccount $GlobalAdminAccount
+
+    $nullReturn = @{
+        GroupID               = $null
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        AllowGiphy            = $null
+        GiphyContentRating    = $null
+        AllowStickersAndMemes = $null
+        AllowCustomMemes      = $null
+    }
+
+
+    Write-Verbose -Message "Getting Team fun settings for $GroupID"
+
+    $teamExists = Get-TeamByGroupID $GroupID
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: Team can only be tied to 1 O365 group so always should return only 1 instance. Its only an issues when filtering by DisplayName

File: Modules/Office365DSC/DSCResources/MSFT_TeamsUser/readme.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Description
+
+This resource used to add new users to a team
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsUser/readme.md]: done

File: Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1 [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    $credsGlobalAdmin = Get-Credential -UserName "TenantAdmin@O365DSC1.onmicrosoft.com" -Message "Global Admin"
+    Node localhost
+    {
+
+        TeamsTeam MyTeam
+        {
+            #GroupId = "f51a3df3-14af-4f52-b22b-30be60ca3fc4"
+            DisplayName = "Sample3"
+            #Description = "Sample"
+            #AccessType = "Private"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1]: Sorry was testing using various scenarios. Fixed

File: Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1 [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            #Alias = "DSCTeam2"
+            GlobalAdminAccount = $credsGlobalAdmin
+            Ensure = "Present"
+        }
+
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1]: Fixed

File: Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1 [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+It is not meant to use as a production baseline.
+#>
+
+Configuration Teams
+{
+    Import-DSCResource -ModuleName Office365DSC
+    $credsGlobalAdmin = Get-Credential -UserName "TenantAdmin@O365DSC1.onmicrosoft.com" -Message "Global Admin"
+    Node localhost
+    {
+
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/Examples/MSFT_TeamsTeam/MSFT_TeamsTeam.ps1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsChannel/MSFT_TeamsChannel.psm1 [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    #Current channel doesnt exist and trying to rename throw an error
+    if (($null -eq $channel) -and $CurrentParameters.ContainsKey("NewDisplayName"))
+    {
+        Write-Verbose "Cannot rename channel $DisplayName , doesnt exist in current Team"
+        throw "Channel named $DisplayName doesnt exist in current Team"
+    }
+
+    if ($null -eq $channel)
+    {
+        Write-Verbose "Failed to get team channels with ID $GroupId and display name of $DisplayName"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsChannel/MSFT_TeamsChannel.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        GroupID               = $GroupID
+        AllowGiphy            = $team.AllowGiphy
+        GiphyContentRating    = $team.GiphyContentRating
+        AllowStickersAndMemes = $team.AllowStickersAndMemes
+        AllowCustomMemes      = $team.AllowCustomMemes
+    }
+
+}
+
+function Set-TargetResource{
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+        $GlobalAdminAccount
+    )
+
+    Write-Verbose -Message "Testing Team fun settings for  $GroupID"
+    $CurrentValues = Get-TargetResource @PSBoundParameters
+
+
+    return Test-Office365DSCParameterState -CurrentValues $CurrentValues `
+        -DesiredValues $PSBoundParameters `
+        -ValuesToCheck @("GiphyContentRating", `
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        [System.Management.Automation.PSCredential]
+        $GlobalAdminAccount
+    )
+
+    Test-TeamsServiceConnection -GlobalAdminAccount $GlobalAdminAccount
+
+    $CurrentParameters = $PSBoundParameters
+    $CurrentParameters.Remove("GlobalAdminAccount")
+    Set-TeamFunSettings @CurrentParameters
+
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        [System.String]
+        $AllowCustomMemes,
+
+        [Parameter(Mandatory = $true)]
+        [System.Management.Automation.PSCredential]
+        $GlobalAdminAccount
+    )
+    Test-TeamsServiceConnection -GlobalAdminAccount $GlobalAdminAccount
+    $result = Get-TargetResource @PSBoundParameters
+    $content = "TeamFunSettings " + (New-GUID).ToString() + "`r`n"
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof [hunk 13]
```diff
@@ -0,0 +1,11 @@
+[ClassVersion("1.0.0.0"), FriendlyName("TeamsFunSettings")]
+class MSFT_TeamsFunSettings : OMI_BaseResource
+{
+    [Key, Description("The group id of the team")] string GroupID;
+    [Write, Description("Allow gilphy in team ")] string AllowGiphy;
+    [Write, Description("Gilphy content rating"),ValueMap{"Strict","Moderate"}, Values{"Strict","Moderate"}] string GiphyContentRating;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof [hunk 14]
```diff
@@ -0,0 +1,11 @@
+[ClassVersion("1.0.0.0"), FriendlyName("TeamsFunSettings")]
+class MSFT_TeamsFunSettings : OMI_BaseResource
+{
+    [Key, Description("The group id of the team")] string GroupID;
+    [Write, Description("Allow gilphy in team ")] string AllowGiphy;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/readme.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Description
+
+This resource used to set Team fun setting on a specific Team
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/readme.md]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+    {
+        Write-Verbose -Message "When Group is passed all other parameters will be ignored."
+    }
+
+    Test-TeamsServiceConnection -GlobalAdminAccount $GlobalAdminAccount
+
+    $nullReturn = @{
+        DisplayName    = $DisplayName
+        Group          = $null
+        GroupId        = $null
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+    Write-Verbose  -Message "Retrieving information about team $($DisplayName) to see if it already exists"
+    $team = Get-TargetResource @PSBoundParameters
+
+    $CurrentParameters = $PSBoundParameters
+    $CurrentParameters.Remove("GlobalAdminAccount")
+    $CurrentParameters.Remove("Ensure")
+
+    if ($Ensure -eq "Present")
+    {
+        if ($team.GroupId)
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+        {
+            ## Can't pass Owner parm into set opertaion and accesstype is called visibility on set 
+            if ($CurrentParameters.ContainsKey("Owner"))
+            {
+                $CurrentParameters.Remove("Owner")
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: Owner can only be set on New-Team not Set-Team

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+            }
+            if ($CurrentParameters.ContainsKey("AccessType"))
+            {
+                $CurrentParameters.Remove("AccessType")
+                $CurrentParameters.Add("Visibility", $AccessType)
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: Its the MS Teams syntax similar to comment on displayName and CurrentDisplayName. On New-Team the attribute is AccessType but on Set-Team its called Visibility

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+
+}
+
+function Test-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
+    param
+    (
+
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.schema.mof [hunk 21]
```diff
@@ -0,0 +1,15 @@
+[ClassVersion("1.0.0.0"), FriendlyName("TeamsTeam")]
+class MSFT_TeamsTeam : OMI_BaseResource
+{
+    [Key, Description("Display Name of the Team")] string DisplayName;
+    [Write, Description("Converts current O365 group to a Team, all other parameters ignore when this is passed into resource.")] string Group;
+    [Write, Description("Description of Team")] string Description;
+    [Write, Description("Team group ID, only used to update existing Team ")] string GroupId;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.schema.mof]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.schema.mof [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+    [Write, Description("Alias of O365 Group associated with Team")] string Alias;
+    [Write, Description("Owner of the Team")] string Owner;
+    [Write, Description("Classification of the Team ")] string Classification;
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.schema.mof]: Its free text

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/readme.md [hunk 23]
```diff
@@ -0,0 +1,3 @@
+# Description
+
+This resource used to create and update new MS Teams
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/readme.md]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsUser/MSFT_TeamsUser.psm1 [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+        [System.Management.Automation.PSCredential]
+        $GlobalAdminAccount
+    )
+
+    Test-TeamsServiceConnection -GlobalAdminAccount $GlobalAdminAccount
+
+    $nullReturn = @{
+        User    = $User
+        Role    = $Role
+        GroupId = $null
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsUser/MSFT_TeamsUser.psm1]: Fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1 [hunk 25]
```diff
@@ -0,0 +1,364 @@
...
+    {
+        $teamGroup = $allGroups | Where-Object {$_.DisplayName -eq $DisplayName}
+    }
+
+    Write-Verbose -Message "Found O365 group $teamGroup"
+    Write-Verbose -Message "Alias = $($teamGroup.Alias) and team accesstype = $($teamGroup.AccessType)"
+
+    return @{
+        DisplayName    = $team.DisplayName
+        Group          = $null
```
[the remark this answers is not in the document]
Comment (desmay) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsTeam/MSFT_TeamsTeam.psm1]: fixed

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (NikCharlebois) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.psm1]: Still not resolved

File: Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof (hunk 13 above)
[the remark this answers is not in the document]
Comment (NikCharlebois) [reply Modules/Office365DSC/DSCResources/MSFT_TeamsFunSettings/MSFT_TeamsFunSettings.schema.mof]: Still don't see the fix

**Pull request merged**